doc/migrations/v1.0.0-v2.0.0.md

```diff
@@ -0,0 +1,245 @@
+<!--Specify versions for migration below-->
+# Migrating to libp2p@2.0 <!-- omit in toc -->
+
+A migration guide for refactoring your application code from libp2p `v1.0.0` to `v2.0.0`.
+
+## Table of Contents <!-- omit in toc -->
+
+- [The autodialer has been removed](#the-autodialer-has-been-removed)
+- ["Transient" connections have been renamed "limited"](#transient-connections-have-been-renamed-limited)
+- [The CustomEvent polyfill has been removed from `@libp2p/interface`](#the-customevent-polyfill-has-been-removed-from-libp2pinterface)
+- [`localPeer` is no longer passed to Connection Encrypters](#localpeer-is-no-longer-passed-to-connection-encrypters)
+- [The `.code` property has been removed from thrown errors, use `.name` instead](#the-code-property-has-been-removed-from-thrown-errors-use-name-instead)
+- [pubsub, identify, circuit-relay v2 and record envelope sealing use `privateKey` component](#pubsub-identify-circuit-relay-v2-and-record-envelope-sealing-use-privatekey-component)
+
+## The autodialer has been removed
+
+The libp2p autodialer was a component that attempted to ensure a minimum number
+of connections were open at any point in time.
+
+When the number of active connections dropped below a configured threshold, it
+did this by searching the peer store for known peers which it then sorted by a
+number of criteria - tag value, recent connections, etc, and would dial them
+until the connection count was above the minimum value.
+
+Opening connections is an expensive operation, and nodes that have unstable
+network conditions will find themselves dialing peers more frequently. This has
+an adverse effect on, for example mobile clients, which would then activate the
+radio more frequently and consume more battery power.
+
+The strategies for peer access vary considerably by protocol. For example
+gossipsub may wish to maintain connections to mesh peers for certain topics in
+order to send/receive messages promptly, but KAD-DHT may just wish to know the
+addresses of a range of peers in the ID space outside of the first KAD-bucket
+and not have a need to be constantly connected to them.
+
+To that end individual protocols can ensure that they have sufficient
+connections for their purposes and libp2p will not try to second-guess their
+rules to apply at a global level.
```

Comment:
Are we planning to encapsulate the auto-dialer, or something else, into a re-usable module that individual protocols can extend to get similar auto-dialer functionality?

Reply:
If individual protocols need to ensure a minimum number of connections to random or arbitrary peers then they should do that.  That way it doesn't burden every other protocol/configuration with potentially unnecessary dials.

That said, we can make something like the autodialer available as a service if we discover that it's necessary.